File: artifacts/api-server/src/routes/catalog.ts
import { Router, type IRouter } from "express";

const router: IRouter = Router();

export interface CatalogGenerateRequest {
  productName?: string;
  craftType?: string;
  materials?: string;
  voiceTranscript?: string;
  imageBase64?: string;
  language?: string;
  materialCost?: number;
  makingHours?: number;
  artisanState?: string;
  testMode?: boolean;
}

export interface CatalogGenerateResponse {
  englishTitle: string;
  hindiTitle: string;
  englishDescription: string;
  hindiDescription: string;
  regionalTitle?: string;
  regionalDescription?: string;
  craftCategory: string;
  materialsDetected: string[];
  careInstructions: string;
  tags: string[];
  geoIndication?: string;
  hsnCode?: string;
  pricing: {
    retail: number;
    wholesale: number;
    export: number;
  };
  fairTradeBreakdown: {
    materialCost: number;
    laborHours: number;
    hourlyWage: number;
    packaging: number;
    wageUpliftPercent: number;
    stateMinWage: number;
  };
}

// State-level statutory minimum wage hourly rates (₹/hr)
const STATE_HOURLY_WAGE: Record<string, number> = {
  UP: 32, // ₹256/day / 8hrs — Uttar Pradesh
  MP: 31, // Madhya Pradesh
  RJ: 30, // Rajasthan
  WB: 32, // West Bengal
  GJ: 35, // Gujarat
  MH: 35, // Maharashtra
  TN: 34, // Tamil Nadu
  KA: 33, // Karnataka
  AP: 31, // Andhra Pradesh
  OR: 29, // Odisha
  JH: 30, // Jharkhand
  BR: 28, // Bihar
  default: 32,
};

/**
 * Comprehensive Indian Craft Knowledge Base
 * Covers 8 major GI-tagged / heritage craft categories with:
 *   - Bilingual titles (English + Hindi)
 *   - Regional language titles (Marathi, Bengali, Tamil, etc.)
 *   - Authentic descriptions referencing real craft clusters
 *   - HSN codes for GeM export compliance
 *   - GI Tag identification
 *   - Care instructions
 */
interface CraftTemplate {
  englishTitle: string;
  hindiTitle: string;
  englishDesc: string;
  hindiDesc: string;
  category: string;
  giTag: string;
  hsnCode: string;
  careInstructions: string;
  materials: string[];
  tags: string[];
}

const CRAFT_TEMPLATES: Record<string, CraftTemplate> = {
  textile: {
    englishTitle: "Handwoven Heritage Textile",
    hindiTitle: "हस्तनिर्मित विरासत वस्त्र",
    englishDesc: "A breathable handloom creation woven in small artisan clusters using traditional pit looms. Dyed with eco-friendly natural colors, carrying the gentle rhythm and story of generations of master weavers. Each thread reflects the cultural identity of India's textile heartland.",
    hindiDesc: "पारंपरिक करघे पर छोटे बैच में बुना गया नरम एवं आरामदायक वस्त्र। प्राकृतिक रंगों से तैयार, हर धागे में भारतीय बुनकरों का हुनर और विरासत झलकती है।",
    category: "Handloom & Heritage Textiles",
    giTag: "Varanasi Silk / Pochampally Ikat Cluster",
    hsnCode: "5007.20.10",
    careInstructions: "Hand wash gently in cold water with mild detergent. Dry flat in shade. Do not wring or bleach.",
    materials: ["Pure Silk", "Natural Dyes", "Gold-Plated Zari Thread"],
    tags: ["#Handloom", "#VaranasiBrocade", "#VocalForLocal", "#HandmadeInIndia", "#FairTrade", "#GITagged"],
  },
  pottery: {
    englishTitle: "Hand-Moulded Terracotta Craft",
    hindiTitle: "हस्तनिर्मित मिट्टी की कलाकृति",
    englishDesc: "Individually sculpted from refined river clay and kiln-fired using age-old pottery methods passed down through Prajapati artisan families. Each piece carries the warmth of earth and fire, perfect for festive celebrations, spiritual ceremonies, and authentic home décor.",
    hindiDesc: "नदी की शुद्ध चिकनी मिट्टी से गढ़ा गया और पारंपरिक भट्टी में पकाया गया। त्योहारों और घर की सजावट के लिए आदर्श, हर कृति में मिट्टी की खुशबू और कारीगर का प्यार बसा है।",
    category: "Terracotta & Clay Craft",
    giTag: "Gorakhpur Terracotta / Molela Terracotta",
    hsnCode: "6912.00.10",
    careInstructions: "Wipe gently with a dry soft cloth. Protect from hard impacts and excessive moisture.",
    materials: ["River Clay", "Natural Pigments", "Kiln-Fired Glaze"],
    tags: ["#Terracotta", "#ClayCraft", "#FestiveDecor", "#EcoFriendly", "#MoSJEArtisan", "#HandmadeInIndia"],
  },
  painting: {
    englishTitle: "Traditional Folk Painting",
    hindiTitle: "पारंपरिक लोक चित्रकला",
    englishDesc: "A stunning hand-painted artwork created using natural mineral pigments and bamboo nibs on handmade Lokta paper. This traditional art form depicts mythological narratives, cosmic geometry, and the sacred connection between nature and humanity — a living tradition spanning over 2,500 years.",
    hindiDesc: "प्राकृतिक खनिज रंगों और बांस की कलम से हस्तनिर्मित कागज़ पर बनाई गई उत्कृष्ट लोक चित्रकला। पौराणिक कथाओं और प्रकृति के सौंदर्य को दर्शाती 2,500 वर्ष पुरानी जीवित परंपरा।",
    category: "Madhubani & Traditional Painting",
    giTag: "Madhubani / Warli / Pattachitra Art",
    hsnCode: "9701.10.00",
    careInstructions: "Frame under UV-protective glass. Avoid direct sunlight and high humidity. Handle with clean dry hands.",
    materials: ["Handmade Paper", "Natural Mineral Pigments", "Bamboo Nib", "Herbal Fixative"],
    tags: ["#MadhubaniArt", "#FolkArt", "#IndianHeritage", "#TribalArt", "#HandPainted", "#LivingTradition"],
  },
  woodcraft: {
    englishTitle: "Intricately Carved Wooden Craft",
    hindiTitle: "बारीक नक्काशीदार लकड़ी का शिल्प",
    englishDesc: "Masterfully hand-carved from sustainably sourced Sheesham rosewood by artisan families of the renowned Saharanpur woodcarving cluster. Features traditional Mughal-inspired jali lattice patterns with brass inlay accents, showcasing centuries of inherited craftsmanship excellence.",
    hindiDesc: "सहारनपुर के प्रसिद्ध लकड़ी नक्काशी परिवारों द्वारा शीशम की लकड़ी पर बारीक हाथ की नक्काशी। मुगल शैली की जाली पैटर्न और पीतल की जड़ाई — सदियों पुरानी कारीगरी की उत्कृष्ट मिसाल।",
    category: "Wood Carving & Lacquerware",
    giTag: "Saharanpur Wood Carving / Channapatna Toys",
    hsnCode: "4420.10.00",
    careInstructions: "Dust with soft dry cloth. Apply furniture wax annually. Keep away from direct heat and moisture.",
    materials: ["Sheesham Rosewood", "Brass Inlay", "Natural Lacquer Polish"],
    tags: ["#WoodCraft", "#Saharanpur", "#JaliArt", "#IndianDecor", "#SustainableWood", "#HandCarved"],
  },
  brass: {
    englishTitle: "Hand-Forged Brass Artware",
    hindiTitle: "हाथ से ढली पीतल की कलाकृति",
    englishDesc: "Skillfully hand-forged using the ancient lost-wax (cire perdue) casting technique by Thathera artisan communities. Each brass vessel is individually hammered, polished, and engraved with traditional geometric motifs — certified food-safe and ideal for Ayurvedic wellness use.",
    hindiDesc: "प्राचीन मोम-गलाई (सिर पर्ड्यू) तकनीक से ठठेरा कारीगरों द्वारा हाथ से ढला गया। हर बर्तन को व्यक्तिगत रूप से पीटा, पॉलिश किया और पारंपरिक ज्यामितीय नक्काशी से सजाया गया है।",
    category: "Brassware & Metal Craft",
    giTag: "Moradabad Brass / Jandiala Guru Thathera",
    hsnCode: "7418.20.00",
    careInstructions: "Clean with lemon and salt paste for natural shine. Rinse with warm water. Dry immediately.",
    materials: ["Pure Brass Alloy", "Copper Base", "Natural Patina Finish"],
    tags: ["#Brassware", "#Thathera", "#MetalCraft", "#AyurvedicVessel", "#IndianHeritage", "#FoodSafe"],
  },
  bluePottery: {
    englishTitle: "Jaipur Blue Pottery",
    hindiTitle: "जयपुर ब्लू पॉटरी",
    englishDesc: "An exquisite piece of Jaipur's iconic blue pottery — crafted without a potter's wheel using a unique Mughal-era technique combining quartz stone dust, powdered glass, and gum. Each piece is hand-painted with cobalt oxide floral motifs and fired to achieve the signature azure blue glaze.",
    hindiDesc: "जयपुर की प्रसिद्ध ब्लू पॉटरी — क्वार्ट्ज पत्थर, कांच पाउडर और गोंद से बिना चाक के बनाई गई। कोबाल्ट ऑक्साइड से हाथ से पेंट की गई फूलों की डिज़ाइन और नीली चमक।",
    category: "Blue Pottery & Ceramic Art",
    giTag: "Jaipur Blue Pottery GI",
    hsnCode: "6913.90.00",
    careInstructions: "Hand wash only. Do not microwave. Avoid thermal shock. Display on cushioned surface.",
    materials: ["Quartz Stone Dust", "Powdered Glass", "Multani Mitti", "Cobalt Oxide"],
    tags: ["#BluePottery", "#Jaipur", "#MughalArt", "#CeramicArt", "#GITagged", "#HomeDecor"],
  },
  jute: {
    englishTitle: "Handcrafted Natural Jute Product",
    hindiTitle: "हस्तनिर्मित जूट उत्पाद",
    englishDesc: "An eco-friendly handcrafted product made from golden jute fiber by skilled artisans of Bengal and Odisha. Woven on traditional handlooms with natural vegetable dyes, this sustainable creation combines rural livelihood empowerment with modern ecological consciousness.",
    hindiDesc: "बंगाल और ओडिशा के कुशल कारीगरों द्वारा सुनहरे पटसन रेशे से बना पर्यावरण अनुकूल उत्पाद। प्राकृतिक वनस्पति रंगों से रंगा और पारंपरिक हथकरघे पर बुना गया।",
    category: "Jute & Eco Craft",
    giTag: "Bengal Jute / Coir Craft",
    hsnCode: "4602.11.00",
    careInstructions: "Spot clean with damp cloth. Air dry in shade. Avoid prolonged water exposure.",
    materials: ["Golden Jute Fiber", "Natural Vegetable Dyes", "Cotton Blend"],
    tags: ["#JuteCraft", "#EcoFriendly", "#Sustainable", "#BengalCraft", "#GreenFashion", "#ZeroWaste"],
  },
  toys: {
    englishTitle: "Traditional Handcrafted Wooden Toy",
    hindiTitle: "पारंपरिक हस्तनिर्मित लकड़ी का खिलौना",
    englishDesc: "A charming handmade wooden toy lathe-turned from sustainably harvested Hale wood and painted with child-safe vegetable lacquer dyes. Created by the legendary Channapatna toy-making artisans of Karnataka — holders of the coveted GI tag and UNESCO recognition.",
    hindiDesc: "कर्नाटक के प्रसिद्ध चन्नापटना खिलौना कारीगरों द्वारा हेल की लकड़ी से बनाया गया सुंदर खिलौना। बच्चों के लिए सुरक्षित वनस्पति लाख के रंगों से पेंट किया गया। GI टैग और यूनेस्को मान्यता प्राप्त।",
    category: "Traditional Toys & Dolls",
    giTag: "Channapatna Toys GI",
    hsnCode: "9503.00.90",
    careInstructions: "Wipe with dry cloth. Non-toxic and child safe. Keep away from water.",
    materials: ["Hale Wood", "Vegetable Lacquer", "Natural Pigments"],
    tags: ["#ChannapatnaToys", "#WoodenToys", "#GITagged", "#ChildSafe", "#IndianToys", "#UNESCO"],
  },
};

/**
 * Detects craft category from input text using keyword matching across
 * product name, craft type, materials, and voice transcript.
 */
function detectCraftCategory(input: string): string {
  const text = input.toLowerCase();
  const checks: [string, RegExp][] = [
    ["textile", /cloth|fabric|dupatta|saree|shawl|cotton|silk|weave|loom|khadi|chikan|bandhani|ikat|patola/],
    ["pottery", /clay|pot|diya|terracotta|ceramic|mud|matka|surahi|kulhad/],
    ["painting", /paint|madhubani|warli|pattachitra|kalamkari|canvas|pigment|miniature|phad/],
    ["woodcraft", /wood|carv|lacquer|furniture|rosewood|sheesham|sandal|teak|jharokha/],
    ["brass", /brass|copper|metal|bell|thathera|bronze|kansa|lota|vessel/],
    ["bluePottery", /blue.?pottery|jaipur.?blue|ceramic.?blue|tile|cobalt|quartz/],
    ["jute", /jute|coir|hemp|basket|bag|eco.?craft|patsun|golden.?fiber/],
    ["toys", /toy|doll|channapatna|lacquer.?toy|puppet|kondapalli|wooden.?toy/],
  ];

  for (const [key, regex] of checks) {
    if (regex.test(text)) return key;
  }
  return "textile"; // Default to handloom textiles
}

router.post("/catalog/generate", async (req, res) => {
  const apiKey = process.env["OPENAI_API_KEY"];
  const {
    productName = "Handcrafted Artisan Product",
    craftType = "Traditional Handicraft",
    materials = "Natural organic materials",
    voiceTranscript = "",
    imageBase64,
    language = "hi",
    materialCost = 350,
    makingHours = 5,
    artisanState = "default",
  } = req.body as CatalogGenerateRequest;

  // Calculate fair-trade pricing
  const stateMinWage = STATE_HOURLY_WAGE[artisanState] ?? STATE_HOURLY_WAGE.default;
  const hourlyWage = Math.round(stateMinWage * 3.5); // Skilled craft wage multiplier
  const laborCost = Math.round(makingHours * hourlyWage);
  const packaging = Math.max(35, Math.round(materialCost * 0.08));
  const baseCost = materialCost + laborCost + packaging;
  const retail = Math.round(baseCost * 2.2);
  const wholesale = Math.round(retail * 0.65);
  const exportPrice = Math.round(retail * 1.35);
  const wageUpliftPercent = Math.round(((hourlyWage - stateMinWage) / stateMinWage) * 100);

  const pricing = { retail, wholesale, export: exportPrice };
  const fairTradeBreakdown = {
    materialCost,
    laborHours: makingHours,
    hourlyWage,
    packaging,
    wageUpliftPercent,
    stateMinWage,
  };

  // 1. If OpenAI API key is present, use GPT-4o-mini Vision for culturally accurate bilingual catalog
  if (apiKey) {
    try {
      const messages: any[] = [
        {
          role: "system",
          content: `You are an expert Indian handicraft curator for the Ministry of Social Justice & Empowerment (MoSJE) and Ministry of Textiles.
Your mission is to help traditional rural artisans sell online.
Create a JSON response with these exact keys:
- englishTitle (compelling, SEO-friendly, 5-8 words)
- hindiTitle (accurate Hindi translation with culturally authentic artisan terms)
- englishDescription (2-3 engaging sentences emphasizing handmade heritage, ethical craft, and texture)
- hindiDescription (authentic Hindi description)
- craftCategory (e.g. Handloom Textiles, Terracotta Pottery, Wood Carving, Brassware, Jute Craft)
- materialsDetected (array of strings)
- careInstructions (concise care tip in English)
- tags (array of 5-8 relevant tags like #VocalForLocal, #HandmadeInIndia)
- geoIndication (if recognizable, e.g. "Jaipur Blue Pottery", "Varanasi Silk", "Gorakhpur Terracotta", "Channapatna Craft", or null)
- hsnCode (4-8 digit HSN code for the craft)
Format as valid JSON only.`,
        },
        {
          role: "user",
          content: [
            ...(imageBase64
              ? [
                  {
                    type: "image_url",
                    image_url: {
                      url: `data:image/jpeg;base64,${imageBase64}`,
                      detail: "low",
                    },
                  },
                ]
              : []),
            {
              type: "text",
              text: `Artisan product: ${productName}
Craft tradition: ${craftType}
Materials used: ${materials}
Artisan voice note description: "${voiceTranscript}"
Language: ${language}
Generate the complete bilingual catalog listing.`,
            },
          ],
        },
      ];

      const response = await fetch("https://api.openai.com/v1/chat/completions", {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: "gpt-4o-mini",
          messages,
          response_format: { type: "json_object" },
          temperature: 0.5,
          max_tokens: 900,
        }),
      });

      if (response.ok) {
        const data = (await response.json()) as any;
        const parsed = JSON.parse(data.choices[0].message.content);
        res.json({
          ...parsed,
          pricing,
          fairTradeBreakdown,
        });
        return;
      }
    } catch (aiErr) {
      req.log.warn({ err: aiErr }, "OpenAI catalog generation failed, using intelligent craft generator");
    }
  }

  // 2. High-quality contextual offline fallback using the full 8-category craft knowledge base
  const inputText = `${productName} ${craftType} ${materials} ${voiceTranscript}`;
  const detectedCategory = detectCraftCategory(inputText);
  const template = CRAFT_TEMPLATES[detectedCategory] ?? CRAFT_TEMPLATES.textile;

  // Customize the title with the product name if it's specific enough
  const hasSpecificName = productName.length > 10 && productName !== "Handcrafted Artisan Product";
  const englishTitle = hasSpecificName
    ? `${productName} — ${template.category}`
    : template.englishTitle;
  const hindiTitle = hasSpecificName
    ? `${productName} — ${template.hindiTitle}`
    : template.hindiTitle;

  // Append voice transcript to description if available
  const voiceNote = voiceTranscript
    ? ` Artisan's personal note: "${voiceTranscript}"`
    : "";

  const catalogResult: CatalogGenerateResponse = {
    englishTitle,
    hindiTitle,
    englishDescription: template.englishDesc + voiceNote,
    hindiDescription: template.hindiDesc,
    craftCategory: template.category,
    materialsDetected: materials ? materials.split(/[,·;]/).map((m) => m.trim()).filter(Boolean) : template.materials,
    careInstructions: template.careInstructions,
    tags: template.tags,
    geoIndication: template.giTag,
    hsnCode: template.hsnCode,
    pricing,
    fairTradeBreakdown,
  };

  res.json(catalogResult);
});

export default router;

File: artifacts/api-server-python/routes/catalog.py
"""
Multilingual AI Auto-Cataloger Route (Python FastAPI)
Uses OpenAI GPT-4o-mini Vision with resilient bilingual handicraft fallback.
"""
import os
import re
import json
from typing import Optional, List, Dict, Any
from fastapi import APIRouter
from pydantic import BaseModel, Field
import httpx

router = APIRouter()

STATE_HOURLY_WAGE: Dict[str, int] = {
    "UP": 32,  # ₹256/day / 8hrs
    "MP": 31,
    "RJ": 30,
    "WB": 32,
    "GJ": 35,
    "MH": 35,
    "default": 32,
}

class CatalogGenerateRequest(BaseModel):
    productName: Optional[str] = "Handcrafted Artisan Product"
    craftType: Optional[str] = "Traditional Handicraft"
    materials: Optional[str] = "Natural organic materials"
    voiceTranscript: Optional[str] = ""
    imageBase64: Optional[str] = None
    language: Optional[str] = "hi"
    materialCost: Optional[int] = 350
    makingHours: Optional[int] = 5
    artisanState: Optional[str] = "default"

class PricingBreakdown(BaseModel):
    retail: int
    wholesale: int
    export: int

class FairTradeBreakdown(BaseModel):
    materialCost: int
    laborHours: int
    hourlyWage: int
    packaging: int

class CatalogGenerateResponse(BaseModel):
    englishTitle: str
    hindiTitle: str
    englishDescription: str
    hindiDescription: str
    regionalTitle: Optional[str] = None
    regionalDescription: Optional[str] = None
    craftCategory: str
    materialsDetected: List[str]
    careInstructions: str
    tags: List[str]
    geoIndication: Optional[str] = None
    pricing: PricingBreakdown
    fairTradeBreakdown: FairTradeBreakdown

@router.post("/catalog/generate", response_model=CatalogGenerateResponse)
async def generate_catalog(payload: CatalogGenerateRequest):
    api_key = os.getenv("OPENAI_API_KEY")
    
    product_name = payload.productName or "Handcrafted Artisan Product"
    craft_type = payload.craftType or "Traditional Handicraft"
    materials = payload.materials or "Natural organic materials"
    voice_transcript = payload.voiceTranscript or ""
    image_b64 = payload.imageBase64
    language = payload.language or "hi"
    material_cost = payload.materialCost or 350
    making_hours = payload.makingHours or 5
    artisan_state = payload.artisanState or "default"

    # Fair-trade pricing calculations
    hourly_wage = STATE_HOURLY_WAGE.get(artisan_state, STATE_HOURLY_WAGE["default"])
    labor_cost = round(making_hours * hourly_wage * 3.5)
    packaging = max(35, round(material_cost * 0.08))
    base_cost = material_cost + labor_cost + packaging
    retail = round(base_cost * 2.2)
    wholesale = round(retail * 0.65)
    export_price = round(retail * 1.35)

    pricing = PricingBreakdown(retail=retail, wholesale=wholesale, export=export_price)
    fair_trade = FairTradeBreakdown(
        materialCost=material_cost,
        laborHours=making_hours,
        hourlyWage=round(hourly_wage * 3.5),
        packaging=packaging,
    )

    # 1. If OpenAI API key present, query GPT-4o-mini Vision
    if api_key:
        try:
            content: List[Dict[str, Any]] = []
            if image_b64:
                content.append({
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/jpeg;base64,{image_b64}",
                        "detail": "low"
                    }
                })
            content.append({
                "type": "text",
                "text": (
                    f"Artisan product: {product_name}\n"
                    f"Craft tradition: {craft_type}\n"
                    f"Materials used: {materials}\n"
                    f'Artisan voice note description: "{voice_transcript}"\n'
                    f"Language: {language}\n"
                    "Generate the complete bilingual catalog listing."
                )
            })

            messages = [
                {
                    "role": "system",
                    "content": (
                        "You are an expert Indian handicraft curator for the Ministry of Social Justice & Empowerment (MoSJE) and Ministry of Textiles.\n"
                        "Your mission is to help traditional rural artisans sell online.\n"
                        "Create a JSON response with these exact keys:\n"
                        "- englishTitle (compelling, SEO-friendly, 5-8 words)\n"
                        "- hindiTitle (accurate Hindi translation with culturally authentic artisan terms)\n"
                        "- englishDescription (2-3 engaging sentences emphasizing handmade heritage, ethical craft, and texture)\n"
                        "- hindiDescription (authentic Hindi description)\n"
                        "- craftCategory (e.g. Handloom Textiles, Terracotta Pottery, Wood Carving, Brassware, Jute Craft)\n"
                        "- materialsDetected (array of strings)\n"
                        "- careInstructions (concise care tip in English)\n"
                        "- tags (array of 5-8 relevant tags like #VocalForLocal, #HandmadeInIndia)\n"
                        "- geoIndication (if recognizable, e.g. 'Jaipur Blue Pottery', 'Varanasi Silk', or null)\n"
                        "Format as valid JSON only."
                    )
                },
                {
                    "role": "user",
                    "content": content
                }
            ]

            async with httpx.AsyncClient(timeout=25.0) as client:
                res = await client.post(
                    "https://api.openai.com/v1/chat/completions",
                    headers={
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": "gpt-4o-mini",
                        "messages": messages,
                        "response_format": {"type": "json_object"},
                        "temperature": 0.5,
                        "max_tokens": 900
                    }
                )
                if res.status_code == 200:
                    data = res.json()
                    parsed = json.loads(data["choices"][0]["message"]["content"])
                    return CatalogGenerateResponse(
                        englishTitle=parsed.get("englishTitle", product_name),
                        hindiTitle=parsed.get("hindiTitle", product_name),
                        englishDescription=parsed.get("englishDescription", ""),
                        hindiDescription=parsed.get("hindiDescription", ""),
                        craftCategory=parsed.get("craftCategory", craft_type),
                        materialsDetected=parsed.get("materialsDetected", [materials]),
                        careInstructions=parsed.get("careInstructions", "Handle with traditional care."),
                        tags=parsed.get("tags", ["#Handmade", "#VocalForLocal"]),
                        geoIndication=parsed.get("geoIndication"),
                        pricing=pricing,
                        fairTradeBreakdown=fair_trade
                    )
        except Exception as e:
            print(f"[Catalog Python] OpenAI generation error: {e}, falling back to intelligent heuristic")

    # 2. Resilient heuristic bilingual fallback (works offline and for live jury demos)
    search_context = f"{product_name} {craft_type} {materials} {voice_transcript}".lower()
    is_textile = bool(re.search(r"cloth|fabric|dupatta|saree|shawl|cotton|silk|weave|loom|khadi", search_context))
    is_pottery = bool(re.search(r"clay|pot|diya|terracotta|ceramic|mud", search_context))

    english_title = product_name
    hindi_title = product_name
    english_desc = "Handcrafted with traditional techniques handed down through generations. Each piece reflects authentic Indian artisan heritage."
    hindi_desc = "पारंपरिक हस्तशिल्प तकनीक से निर्मित, हर कृति में भारतीय कारीगरी और सांस्कृतिक विरासत की अनूठी झलक है।"
    craft_cat = craft_type
    gi_tag = None

    if is_textile:
        english_title = product_name if "Dupatta" in product_name else f"Handwoven {product_name} with Natural Dyes"
        hindi_title = "हाथ से बुना हुआ पारंपरिक दुपट्टा"
        english_desc = f"A soft, breathable handloom creation woven in small artisan clusters. Dyed with eco-friendly natural colors, carrying the gentle rhythm and story of traditional pit looms. {f'Artisan note: {voice_transcript}' if voice_transcript else ''}"
        hindi_desc = "पारंपरिक करघे पर छोटे बैच में बुना गया नरम एवं आरामदायक वस्त्र। प्राकृतिक रंगों से तैयार, हर धागे में भारतीय बुनकरों का हुनर झलकता है।"
        craft_cat = "Handloom & Heritage Textiles"
        gi_tag = "Varanasi / Pochampally Handloom Cluster"
    elif is_pottery:
        english_title = f"Hand-moulded Terracotta {product_name}"
        hindi_title = "हस्तनिर्मित नक्काशीदार मिट्टी की कलाकृति"
        english_desc = f"Individually sculpted from refined river clay and kiln-fired using age-old pottery methods. Perfect for festive celebrations and home decor. {f'Artisan note: {voice_transcript}' if voice_transcript else ''}"
        hindi_desc = "नदी की शुद्ध चिकनी मिट्टी से गढ़ा गया और पारंपरिक भट्टी में पकाया गया सुंदर दीया। त्योहारों और घर की सजावट के लिए आदर्श।"
        craft_cat = "Terracotta & Clay Craft"
        gi_tag = "Gorakhpur Terracotta Craft"

    detected_materials = [m.strip() for m in re.split(r"[,·]", materials)] if materials else ["Natural Craft Material"]

    return CatalogGenerateResponse(
        englishTitle=english_title,
        hindiTitle=hindi_title,
        englishDescription=english_desc,
        hindiDescription=hindi_desc,
        craftCategory=craft_cat,
        materialsDetected=detected_materials,
        careInstructions="Hand wash gently in cold water." if is_textile else "Wipe gently with a soft dry cloth.",
        tags=[
            "#VocalForLocal",
            "#HandmadeInIndia",
            "#MoSJEArtisan",
            "#HeritageCraft",
            "#FairTrade",
            "#AtmanirbharBharat"
        ],
        geoIndication=gi_tag,
        pricing=pricing,
        fairTradeBreakdown=fair_trade
    )
